File: 1061. Lexicographically Smallest Equivalent String/disjoint.py
class Solution:
    def smallestEquivalentString(self, s1: str, s2: str, baseStr: str) -> str:
        bag = {}

        def find(alpha):
            bag.setdefault(alpha, alpha)
            if bag[alpha] != alpha:
                bag[alpha] = find(bag[alpha])

            return bag[alpha]

        def union(alpha1, alpha2):
            root1 = find(alpha1)
            root2 = find(alpha2)

            if root1 >= root2:
                bag[root1] = root2
            else:
                bag[root2] = root1

        for index in range(len(s1)):
            union(s1[index], s2[index])

        res = ""

        for alpha in baseStr:
            res += find(alpha)

        return res

File: 1061. Lexicographically Smallest Equivalent String/disjoint.cpp
class Solution{
public:
    string smallestEquivalentString(string s1, string s2, string baseStr) {
        
        memset(bag, -1, sizeof(bag));

        for (auto i = 0; i < s1.size() ; i++)
            Union(s1[i] - 'a', s2[i] - 'a');

        for (auto i = 0; i < baseStr.size(); i++)
            baseStr[i] = Find(baseStr[i] - 'a') + 'a';

        return baseStr;
    }
private:
    int bag[26];

    int Find(int x) {
        if (bag[x] == -1)
            return x;
        return bag[x] = Find(bag[x]);
    }

    void Union(int x, int y) {
        int fx = Find(x);
        int fy = Find(y);
        if (fx != fy)
            if (fx < fy)
                bag[fy] = fx;
            else
                bag[fx] = fy;
    }
};